### PR TITLE
authz/azuredevops: Check enforcePermissions in NewAuthzProviders

### LINKED ISSUE #47165: Implement user permissions sync for Azure code host
**Blocked by**

- [ ] #47163.

**Tasks**

- [x] ~~Use the list organization API to list all organizations~~
- [x] Use the list repositories API to list all repos under an organizations #48209 
- [x] Populate permissions in the user permissions table #48209 
- [x] Implement ValidationCheck #48465 

/cc @ryphil

### PR DESCRIPTION
**Pull request by indradhanush**
We do not want to run permission sync for orgs and projects that are declared in code host connections without an explicit `enforcePermissions: true`. Skip them instead.

Follow up to https://github.com/sourcegraph/sourcegraph/pull/48209#discussion_r1122990998.

Fixes #47165.

## Note to reviewers

The code change itself is small but important. Most of the added and removed lines is new tests and some refactoring of the tests.

I've added inline comments to explain the changes which should make it reviewing this more approachable.

## Test plan
- Added tests
- Build should pass

### CODE REVIEW

File: enterprise/internal/authz/azuredevops/provider_test.go [hunk 1]
```diff
@@ -70,19 +208,14 @@ func TestProvider_FetchUserPerms(t *testing.T) {
...
-		mockServer       *httptest.Server
+		connection *schema.AzureDevOpsConnection
+		account    *extsvc.Account
+		mockServer *httptest.Server
 	}
 
 	type output struct {
-		invalidConnections []string
-		problems           []string
-		warnings           []string
```
Comment (indradhanush) [inline enterprise/internal/authz/azuredevops/provider_test.go:82]: Now we test for these in the newly added top level test `TestProvider_NewAuthzProviders`. Previously we were testing implicitly for `NewAuthzProviders` in this test itself, but it's better to have a separate test.

File: enterprise/internal/authz/azuredevops/provider_test.go [hunk 2]
```diff
@@ -91,24 +224,10 @@ func TestProvider_FetchUserPerms(t *testing.T) {
...
-				mockCheckFeature: func(_ licensing.Feature) error {
-					return errors.New("not allowed")
-				},
-				connection: &schema.AzureDevOpsConnection{},
-			},
-			output: output{
-				invalidConnections: []string{"azuredevops"},
-				problems:           []string{"not allowed"},
-			},
-		},
```
Comment (indradhanush) [inline enterprise/internal/authz/azuredevops/provider_test.go:106]: Moved up to the new test.

File: enterprise/internal/authz/azuredevops/provider_test.go [hunk 3]
```diff
@@ -70,19 +208,14 @@ func TestProvider_FetchUserPerms(t *testing.T) {
 	}
 
 	type input struct {
-		mockCheckFeature func(licensing.Feature) error
```
Comment (indradhanush) [inline enterprise/internal/authz/azuredevops/provider_test.go:73]: All the unit tests in this want the licensing check to pass. Removed it as an input and explicitly setting it below at the start of testing.

File: enterprise/internal/authz/azuredevops/provider_test.go [hunk 4]
```diff
@@ -70,19 +208,14 @@ func TestProvider_FetchUserPerms(t *testing.T) {
...
+		connection *schema.AzureDevOpsConnection
+		account    *extsvc.Account
+		mockServer *httptest.Server
 	}
 
 	type output struct {
-		invalidConnections []string
-		problems           []string
-		warnings           []string
-		providers          []authz.Provider
```
Comment (indradhanush) [inline enterprise/internal/authz/azuredevops/provider_test.go:83]: This was the same for all the tests - `expectedProvidrers`. Stopped passing it as an arg and instead using `expectedProviders` directly.

File: enterprise/internal/authz/azuredevops/provider_test.go [hunk 5]
```diff
@@ -215,10 +329,10 @@ func TestProvider_FetchUserPerms(t *testing.T) {
 			name:  "auth provider config with both orgs and projects",
 			setup: setup,
 			input: input{
-				mockCheckFeature: passLicensingCheck,
 				connection: &schema.AzureDevOpsConnection{
-					Orgs:     []string{"solarsystem"},
-					Projects: []string{"solar/system"},
+					EnforcePermissions: true,
```
Comment (indradhanush) [inline enterprise/internal/authz/azuredevops/provider_test.go:333]: Needed to start setting this for the tests to pass as a result of the new change in this PR.

File: enterprise/internal/authz/azuredevops/provider_test.go [hunk 6]
```diff
@@ -329,36 +442,23 @@ func TestProvider_FetchUserPerms(t *testing.T) {
 				conf.Mock(nil)
 			})
 
-			licensing.MockCheckFeature = tc.mockCheckFeature
```
Comment (indradhanush) [inline enterprise/internal/authz/azuredevops/provider_test.go:332]: Set this once for all the unit tests here.

File: enterprise/internal/authz/azuredevops/provider.go [hunk 7]
```diff
@@ -26,15 +26,15 @@ var mockServerURL string
 func NewAuthzProviders(db database.DB, conns []*types.AzureDevOpsConnection) *authztypes.ProviderInitResult {
 	orgs, projects := map[string]struct{}{}, map[string]struct{}{}
 
-	initResults := &authztypes.ProviderInitResult{}
-
-	if len(conns) == 0 {
-		return initResults
-	}
+	validAzureDevOpsConnections := []*types.AzureDevOpsConnection{}
```
Comment (kopancek) [inline enterprise/internal/authz/azuredevops/provider.go:29]: Naming is hard, maybe call this `authorizedConnections` instead? Or `connectionsWithPermissions`. Valid evokes in me as if some other config was invalid :)

**Review by kopancek — APPROVED** [review]
Looks good to me

File: enterprise/internal/authz/azuredevops/provider.go (hunk 7 above)
Comment (indradhanush) [reply enterprise/internal/authz/azuredevops/provider.go:29]: I like that name. Going ahead with `authorizedConnections`. 👍

**Pull request merged**